src/inference/ess.jl

```diff
@@ -112,7 +112,9 @@ function Base.rand(rng::Random.AbstractRNG, p::ESSPrior)
     sampler = p.sampler
     varinfo = p.varinfo
     vns = _getvns(varinfo, sampler)
-    set_flag!(varinfo, vns[1][1], "del")
+    for vn in Iterators.flatten(values(vns))
+        set_flag!(varinfo, vn, "del")
+    end
```

Comment:
It works sometimes because in `assume` we only check if the flag is set for `vns[1]`, but that is only for "sub-symbols". This is the line I'm referring to: https://github.com/TuringLang/DynamicPPL.jl/blob/9083299db3f623136895cae80ef5f10d7fcf8d2c/src/context_implementations.jl#L268. But this won't work if we have more than one key in `vi.metadata` or if `tilde_assume` and others are called with a varname subsumed e.g. `m[2]`.

And this won't be an issue once we have a clear separation between sampling and evaluation. These sorts of bugs show up soooo often (and I agree, it's super-confusing), so looking forward to not having those:)